### PR TITLE
Better waterworld.

### PR DESCRIPTION
**Pull request by alexhliu**
Switch water-world to use the blockpalette water which allows for changing it's material properties.
Removed `Water.OCEAN_WATER` since it was not being used outside of water-world.
image

### CODE REVIEW

File: chunky/src/java/se/llbit/chunky/renderer/scene/PreviewRayTracer.java [hunk 1]
```diff
@@ -110,18 +110,23 @@ public static boolean nextIntersection(Scene scene, Ray ray) {
   }
 
   private static boolean waterIntersection(Scene scene, Ray ray) {
+    double t = (scene.getEffectiveWaterPlaneHeight() - ray.o.y - scene.origin.y) / ray.d.y;
+    if (scene.waterPlaneClip) {
```
Comment (ShirleyNekoDev) [inline chunky/src/java/se/llbit/chunky/renderer/scene/PreviewRayTracer.java:114]: Could this use a getter? 🙈

File: chunky/src/java/se/llbit/chunky/renderer/scene/Scene.java [hunk 2]
```diff
@@ -244,6 +244,7 @@
   protected boolean waterPlaneEnabled = false;
   protected double waterPlaneHeight = World.SEA_LEVEL;
   protected boolean waterPlaneOffsetEnabled = true;
+  protected boolean waterPlaneClip = true;
```
Comment (ShirleyNekoDev) [inline chunky/src/java/se/llbit/chunky/renderer/scene/Scene.java:247]: I would maybe rename it to waterPlaneChunkClip to make it clear what is getting clipped.

File: chunky/src/java/se/llbit/chunky/renderer/scene/PreviewRayTracer.java (hunk 1 above)
Comment (leMaik) [reply chunky/src/java/se/llbit/chunky/renderer/scene/PreviewRayTracer.java:114]: This calculates `t` if `ray.d.y == 0`, which is not needed. I guess that never really happens a lot though.

File: chunky/src/res/se/llbit/chunky/ui/render/WaterTab.fxml [hunk 3]
```diff
@@ -31,6 +31,7 @@
           </padding>
           <DoubleAdjuster fx:id="waterPlaneHeight" />
           <CheckBox fx:id="waterPlaneOffsetEnabled" mnemonicParsing="false" text="Lower water by minecraft offset (default)" />
+          <CheckBox fx:id="waterPlaneClip" mnemonicParsing="false" text="Clip the water plane in loaded chunks" />
```
Comment (leMaik) [inline chunky/src/res/se/llbit/chunky/ui/render/WaterTab.fxml:34]: The word "clip" confused me here. How about

```suggestion
          <CheckBox fx:id="waterPlaneClip" mnemonicParsing="false" text="Hide the water plane in loaded chunks" />
```

File: chunky/src/java/se/llbit/chunky/block/Water.java [hunk 4]
```diff
@@ -21,9 +21,6 @@
   public static final Water INSTANCE = new Water(0);
 
   public static final double TOP_BLOCK_GAP = 0.125;
-  
-  // Used only as starting material when camera is submerged.
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/block/Water.java:25]: I wonder if this was important. Like is there something that is broken now, e.g. if the camera is submerged?

File: chunky/src/java/se/llbit/chunky/renderer/scene/Scene.java [hunk 5]
```diff
@@ -2468,6 +2479,17 @@ public void setBufferFinalization(boolean value) {
     }
   }
 
+  /**
+   * Query if a position is loaded.
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/renderer/scene/Scene.java:2483]:
```suggestion
   * Query if a chunk is loaded.
```

**Review by leMaik — APPROVED** [review]

**Comment by leMaik** [comment]
There is a off-by-one error somewhere. This only happens at two sides of the loaded chunks (negative x/z):

image

**Comment by leMaik** [comment]
In a perfect world we could take the water world mode into account and not make the water change height here… but that would require re-loading the chunks when toggling water world mode.

image

**Pull request merged**